### DISCUSSION TITLE
How to achieve stateless build and custom api url.

### DISCUSSION THREAD

**Discussion by paeolo** [vercel/next.js#12413] [Help]
I would like to use Next with an external API. I want to inject the API_URL into the Next Application. 

One way to achieve this is to use publicRuntimeConfig as described here : https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration.

In my next.config.js I added the following code:
>   publicRuntimeConfig: { API_URL: process.env.API_URL }

And in my __app.js:
>  const { publicRuntimeConfig } = getConfig()
> OpenAPI.options.url = publicRuntimeConfig.API_URL

It works well, besides If you do that, your build is not stateless anymore. Indeed, static pages are rendered at build time with the API_URL that was in your environment variable making your build statefull. Some platform like Heroku proposes interesting features like promoting a build from staging to production, but it requires a stateless build. 

Now suppose you want a stateless build, but you still want to inject the API_URL into the client, even for statically rendered pages. Under the hood, when you run your NextJS server it would load your environment variable, and serve it to the client along with the pages, and the pages will use that even they were statically rendered.

Is there any way to achieve this ?

**Comment by jokester** [comment]
Would plain environment variable ( https://nextjs.org/docs/api-reference/next.config.js/environment-variables ) work in this case?

**Reply by paeolo** [reply]
I dont think so, from the very same page:
> Next.js will replace process.env.customKey with 'my-value' at build time.
